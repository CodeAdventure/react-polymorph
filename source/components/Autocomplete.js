// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';


// external libraries
import createRef from 'create-react-ref/lib/createRef';
import _ from 'lodash';

// interal components
import { GlobalListeners } from './HOC/GlobalListeners';
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';
import { composeFunctions } from '../utils/props';

import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  error: string,
  invalidCharsRegex: RegExp,
  isOpeningUpward: boolean,
  label: string | Element<any>,
  maxSelections: number,
  maxVisibleOptions: number,
  multipleSameSelections: boolean,
  onChange: Function,
  options: Array<any>,
  preselectedOptions: Array<any>,
  placeholder: string,
  renderSelections: Function,
  renderOptions: Function,
  skin: ComponentType<any>,
  sortAlphabetically: boolean,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  inputValue: string,
  error: string,
  selectedOptions: Array<any>,
  filteredOptions: Array<any>,
  isOpen: boolean,
  composedTheme: Object
};

class AutocompleteBase extends Component<Props, State> {
  // declare ref types
  rootElement: ?Element<any>;
  inputElement: ?Element<'input'>;
  suggestionsElement: ?Element<any>;
  optionsElement: ?Element<any>;

  // define static properties
  static displayName = 'Autocomplete';
  static defaultProps = {
    error: null,
    invalidCharsRegex: /[^a-zA-Z0-9]/g, // only allow letters and numbers by default
    isOpeningUpward: false,
    maxVisibleOptions: 10, // max number of visible options
    multipleSameSelections: true, // if true then same word can be selected multiple times
    options: [],
    sortAlphabetically: true, // options are sorted alphabetically by default
    theme: null,
    themeId: IDENTIFIERS.AUTOCOMPLETE,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    // define refs
    this.rootElement = createRef();
    this.inputElement = createRef();
    this.suggestionsElement = createRef();
    this.optionsElement = createRef();

    const {
      context,
      themeId,
      theme,
      themeOverrides,
      sortAlphabetically,
      options,
      preselectedOptions
    } = props;

    this.state = {
      inputValue: '',
      error: '',
      selectedOptions: preselectedOptions || [],
      filteredOptions:
        sortAlphabetically && options ? options.sort() : options || [],
      isOpen: false,
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  clear = () => this._removeOptions();

  focus = () => this.handleAutocompleteClick();

  open = () => this.setState({ isOpen: true });

  close = () => this.setState({ isOpen: false });

  toggleOpen = () => this.setState({ isOpen: !this.state.isOpen });

  handleAutocompleteClick = () => {
    const { inputElement } = this;
    if (inputElement && inputElement.current) {
      inputElement.current.focus();
    }
    // toggle options open/closed
    this.toggleOpen();
  };

  onKeyDown = (event: SyntheticKeyboardEvent<>) => {

    if ( // Check for backspace in order to delete the last selected option
      event.keyCode === 8 &&
      !event.target.value &&
      this.state.selectedOptions.length
    ) {
      // Remove last selected option
      this.removeOption(this.state.selectedOptions.length - 1, event);
    } else if (event.keyCode === 27) { // ESCAPE: prevents Modal from closing
      event.stopPropagation();
    } else if (event.keyCode === 13) { // ENTER: opens Options
      this.open();
    }
  };

  // onChange handler for input element in AutocompleteSkin
  handleInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this._setInputValue(event.target.value);
  };

  // passed to Options onChange handler in AutocompleteSkin
  handleChange = (option: any, event: SyntheticEvent<>) => {
    this.updateSelectedOptions(event, option);
  };

  updateSelectedOptions = (
    event: SyntheticEvent<>,
    selectedOption: any = null
  ) => {
    const canMoreOptionsBeSelected =
      this.state.selectedOptions.length < this.props.maxSelections;
    const areFilteredOptionsAvailable =
      this.state.filteredOptions && this.state.filteredOptions.length > 0;

    if (
      !this.props.maxSelections ||
      (canMoreOptionsBeSelected && areFilteredOptionsAvailable)
    ) {
      if (!selectedOption) return;
      const option = selectedOption.trim();
      const optionCanBeSelected =
        (this.state.selectedOptions.indexOf(option) < 0 &&
          !this.props.multipleSameSelections) ||
        this.props.multipleSameSelections;

      if (option && optionCanBeSelected && this.state.isOpen) {
        const selectedOptions = _.concat(this.state.selectedOptions, option);
        this.selectionChanged(selectedOptions, event);
        this.setState({ selectedOptions, isOpen: false });
      }
    }

    this._setInputValue('');
  };

  removeOption = (index: number, event: SyntheticEvent<>) => {
    const selectedOptions = this.state.selectedOptions;
    _.pullAt(selectedOptions, index);
    this.selectionChanged(selectedOptions, event);
    this.setState({ selectedOptions });
  };

  selectionChanged = (
    selectedOptions: Array<any>,
    event: SyntheticEvent<any>
  ) => {
    if (this.props.onChange) this.props.onChange(selectedOptions, event);
  };

  // returns an object containing props, theme, and method handlers
  // associated with rendering this.state.selectedOptions, the user can call
  // this in the body of the renderSelections function
  getSelectionProps = ({
    removeSelection
  }: { removeSelection: Function } = {}) => {
    const { themeId } = this.props;
    const { inputValue, isOpen, selectedOptions, composedTheme } = this.state;
    return {
      inputValue,
      isOpen,
      selectedOptions,
      theme: composedTheme[themeId],
      removeSelection: (index: number, event: SyntheticEvent<>) =>
        // the user's custom removeSelection event handler is composed with
        // the internal functionality of Autocomplete (this.removeOption)
        composeFunctions(removeSelection, this.removeOption)(index, event)
    };
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      context,
      invalidCharsRegex,
      multipleSameSelections,
      sortAlphabetically,
      skin: AutocompleteSkin,
      theme,
      themeOverrides,
      onChange,
      error,
      ...rest
    } = this.props;

    return (
      <GlobalListeners
        optionsIsOpen={this.state.isOpen}
        optionsRef={this.optionsElement}
        rootRef={this.rootElement}
        toggleOpen={this.toggleOpen}
      >
        {() => (
          <AutocompleteSkin
            error={error || this.state.error}
            filteredOptions={this.state.filteredOptions}
            getSelectionProps={this.getSelectionProps}
            handleAutocompleteClick={this.handleAutocompleteClick}
            handleChange={this.handleChange}
            handleInputChange={this.handleInputChange}
            inputRef={this.inputElement}
            inputValue={this.state.inputValue}
            isOpen={this.state.isOpen}
            onKeyDown={this.onKeyDown}
            optionsRef={this.optionsElement}
            removeOption={this.removeOption}
            rootRef={this.rootElement}
            selectedOptions={this.state.selectedOptions}
            suggestionsRef={this.suggestionsElement}
            theme={this.state.composedTheme}
            toggleOpen={this.toggleOpen}
            {...rest}
          />
        )}
      </GlobalListeners>
    );
  }

  // ======== PRIVATE METHOD ==========

  _removeOptions = () => {
    const { onChange } = this.props;
    onChange ? onChange([]) : null;
    this.setState({ selectedOptions: [], inputValue: '' });
  };

  _filterOptions = (value: string) => {
    let filteredOptions = [];

    if (value !== '') {
      _.some(this.props.options, (option) => {
        if (_.startsWith(option, value)) {
          filteredOptions.push(option);
        }
      });
    } else {
      filteredOptions = this.props.options;
    }

    return filteredOptions;
  };

  _filterInvalidChars = (value: string) => {
    let filteredValue = '';

    if (this.props.invalidCharsRegex.test(value)) {
      filteredValue = value.replace(this.props.invalidCharsRegex, '');
    } else {
      filteredValue = value;
    }

    return filteredValue;
  };

  _setInputValue = (value: string) => {
    const filteredValue = this._filterInvalidChars(value);
    const filteredOptions = this._filterOptions(filteredValue);
    this.setState({
      isOpen: true,
      inputValue: filteredValue,
      filteredOptions
    });
  }
}

export const Autocomplete = withTheme(AutocompleteBase);
