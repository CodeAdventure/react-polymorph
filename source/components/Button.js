// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal components
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  label: string | Element<any>,
  loading: boolean,
  onClick: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object // custom css/scss from user that adheres to component's theme API
};

type State = {
  composedTheme: Object
};

class ButtonBase extends Component<Props, State> {
  // define static properties
  static displayName = 'Button';
  static defaultProps = {
    disabled: false,
    loading: false,
    theme: null,
    themeId: IDENTIFIERS.BUTTON,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: ButtonSkin,
      theme,
      themeOverrides,
      context,
      ...rest
    } = this.props;

    return <ButtonSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export const Button = withTheme(ButtonBase);
