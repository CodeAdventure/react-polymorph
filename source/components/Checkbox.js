import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import utility functions
import { StringOrElement } from '../utils/props.js';
import composeTheme from '../utils/composeTheme.js';

// import the Checkbox component's theme API
import { checkboxThemeAPI } from '../themes/API/checkbox.js';

export default class Checkbox extends Component {
  static propTypes = {
    checked: PropTypes.bool,
    label: StringOrElement,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    skin: PropTypes.func.isRequired,
    theme: PropTypes.object,
    themeOverrides: PropTypes.object,
    themeAPI: PropTypes.object
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    theme: {},
    themeOverrides: {},
    themeAPI: { ...checkboxThemeAPI }
  };

  render() {
    const {
      skin: CheckboxSkin,
      theme,
      themeOverrides,
      themeAPI,
      ...rest
    } = this.props;

    const composedTheme = composeTheme(theme, themeOverrides, themeAPI);

    return <CheckboxSkin theme={composedTheme} {...rest} />;
  }
}
