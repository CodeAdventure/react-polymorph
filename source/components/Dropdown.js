// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';
import createRef from 'create-react-ref/lib/createRef';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

type Props = {
  activeItem: any,
  className?: string,
  clickToOpen?: boolean,
  context: ThemeContextProp,
  isOpen?: boolean,
  items: Array<any>,
  label: string | Element<any>,
  noArrow?: boolean,
  onItemSelected?: Function,
  skin?: ComponentType<any>,
  theme: ?Object,
  themeId: string,
  themeOverrides: Object,
};

type State = {
  composedTheme: Object,
  isOpen: boolean,
  isMouseOverItems: boolean,
  isMouseOverRoot: boolean,
};

class DropdownBase extends Component<Props, State> {
  // declare ref types
  rootElement: ?Element<*>;

  // define static properties
  static displayName = 'Dropdown';
  static defaultProps = {
    context: createEmptyContext(),
    clickToOpen: false,
    noArrow: false,
    theme: null,
    themeOverrides: {},
    themeId: IDENTIFIERS.DROPDOWN,
  };

  constructor(props: Props) {
    super(props);

    // define ref
    this.rootElement = createRef();

    const { context, themeId, theme, themeOverrides } = props;
    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      isOpen: false,
      isMouseOverItems: false,
      isMouseOverRoot: false,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    didThemePropsChange(this.props, nextProps, this.setState.bind(this));
  }

  // ========= PUBLIC SKIN API =========

  toggleOpen = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  // ========= PRIVATE SKIN API =========

  _setMouseOverItems = (isMouseOverItems: boolean) => {
    this.setState({ isMouseOverItems });
  };

  _setMouseOverRoot = (isMouseOverRoot: boolean) => {
    this.setState({ isMouseOverRoot });
  };

  _onItemSelected = (item) => {
    const { onItemSelected } = this.props;
    this._setMouseOverRoot(false);
    this._setMouseOverItems(false);
    this.setState({ isOpen: false });
    if (onItemSelected) {
      onItemSelected(item);
    }
  };

  _onLabelClick = () => {
    this.toggleOpen();
  };

  render() {
    const {
      clickToOpen,
      context,
      isOpen,
      onItemSelected,
      skin,
      theme,
      themeOverrides,
      ...rest
    } = this.props;

    const DropdownSkin = skin || context.skins[IDENTIFIERS.DROPDOWN];
    const { isMouseOverItems, isMouseOverRoot } = this.state;
    const isOpenBecauseOfHover = clickToOpen ? false : isMouseOverItems || isMouseOverRoot;

    return (
      <DropdownSkin
        isOpen={isOpen || this.state.isOpen || isOpenBecauseOfHover}
        onLabelClick={this._onLabelClick}
        onItemSelected={this._onItemSelected}
        rootRef={this.rootElement}
        theme={this.state.composedTheme}
        setMouseOverRoot={this._setMouseOverRoot}
        setMouseOverItems={this._setMouseOverItems}
        {...rest}
      />
    );
  }
}

export const Dropdown = withTheme(DropdownBase);
