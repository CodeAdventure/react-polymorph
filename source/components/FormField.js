// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal components
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  error: string,
  inputRef: Object,
  label: string | Element<any>,
  render: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  error: string,
  composedTheme: Object
};

class FormFieldBase extends Component<Props, State> {
  // define static properties
  static displayName = 'FormField';
  static defaultProps = {
    disabled: false,
    theme: null,
    themeId: IDENTIFIERS.FORM_FIELD,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      error: '',
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  setError = (error: string) => this.setState({ error });

  focusChild = () => {
    const { inputRef } = this.props;
    if (!inputRef.current) return;
    inputRef.current.focus();
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: FormFieldSkin,
      theme,
      themeOverrides,
      error,
      context,
      inputRef,
      ...rest
    } = this.props;

    return (
      <FormFieldSkin
        error={error || this.state.error}
        setError={this.setError}
        theme={this.state.composedTheme}
        focusChild={this.focusChild}
        {...rest}
      />
    );
  }
}

export const FormField = withTheme(FormFieldBase);
