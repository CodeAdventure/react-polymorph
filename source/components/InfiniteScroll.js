// @flow
import React, { Component } from 'react';
// $FlowFixMe
import createRef from 'create-react-ref/lib/createRef';

// internal components
import { withTheme } from './HOC/withTheme';

// utilities
import { composeTheme, addThemeId } from '../utils/themes';

// constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  fetchData: Function,
  renderItems: Function,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object,
  threshold: number
};

type State = {
  data: Object | Array,
  error: string | Element<*>,
  isLoading: boolean,
  hasMoreData: boolean
};

class InfiniteScrollBase extends Component<Props, State> {
  // declare ref types
  scrollContainer: ?Element<'div'>;

  // define static properties
  static displayName = 'InfiniteScroll';
  static defaultProps = {
    fetchData() {},
    theme: null,
    themeId: IDENTIFIERS.INFINITE_SCROLL,
    themeOverrides: {},
    threshold: 250
  };

  constructor(props: Props) {
    super(props);
    const { context, themeId, theme, themeOverrides } = props;

    // refs
    this.scrollContainer = createRef();

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      data: [],
      error: false,
      isLoading: false,
      hasMoreData: true
    };
  }

  componentWillMount() {
    this.handleFetchData();
  }

  componentDidMount() {
    const { scrollContainer } = this;
    scrollContainer.current.addEventListener('scroll', this.handleScroll);
  }

  _isFunction = (renderProp: ?Function) => (renderProp && typeof renderProp === 'function');

  setError = (error: string | Element<*>) => this.setState({ error });

  handleFetchData = () => this.props.fetchData(this.setState.bind(this));

  checkForScrollBottom = () => {
    const { scrollContainer, props: { threshold } } = this;
    const { offsetHeight, scrollTop, scrollHeight } = scrollContainer.current;

    if (offsetHeight + scrollTop >= scrollHeight - threshold) {
      return this.handleFetchData();
    }
  };

  handleScroll = () => {
    const { error, isLoading, hasMoreData } = this.state;

    // return early for error, loading, or lack of future data
    if (error || isLoading || !hasMoreData) { return; }
    return this.checkForScrollBottom();
  };

  render() {
    const { renderItems, themeId } = this.props;
    const { composedTheme, ...restState } = this.state;
    const theme = composedTheme[themeId];

    if (!this._isFunction(renderItems)) { return null; }
    return (
      <div ref={this.scrollContainer} className={theme.root}>
        {renderItems({ ...restState, theme })}
      </div>
    );
  }
}

export const InfiniteScroll = withTheme(InfiniteScrollBase);
