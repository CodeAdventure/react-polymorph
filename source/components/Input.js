// @flow
import React, { Component } from 'react';
// $FlowFixMe
import type { ComponentType, Element, SyntheticInputEvent } from 'react';

// external libraries
import createRef from 'create-react-ref/lib/createRef';
import { isString, flow } from 'lodash';

// internal components
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  autoFocus: boolean,
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  error: string,
  label: string | Element<any>,
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  maxLength: number,
  minLength: number,
  onKeyPress: Function,
  placeholder: string,
  readOnly: boolean,
  setError: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object,
  value: string
};

type State = {
  error: string,
  composedTheme: Object
};

class InputBase extends Component<Props, State> {
  // declare ref types
  inputElement: Element<'input'>;

  // define static properties
  static displayName = 'Input';
  static defaultProps = {
    autoFocus: false,
    error: '',
    readOnly: false,
    theme: null,
    themeId: IDENTIFIERS.INPUT,
    themeOverrides: {},
    value: ''
  };

  constructor(props: Props) {
    super(props);

    // define ref
    this.inputElement = createRef();

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      error: ''
    };
  }

  componentDidMount() {
    if (this.props.autoFocus) this.focus();
  }

  onChange = (event: SyntheticInputEvent<Element<'input'>>) => {
    const { onChange, disabled } = this.props;
    if (disabled) return;

    if (onChange) onChange(this._processValue(event.target.value), event);
  };

  focus = () => {
    const { inputElement } = this;
    if (!inputElement.current) return;
    inputElement.current.focus();
  };

  _setError = (error: string) => {
    const { setError } = this.props;

    // checks for setError func from FormField component
    // if this Input instance is being used within the render function
    // of a FormField instance, the error field within FormField's local state
    // will be set
    if (setError) setError(error);
    this.setState({ error });
  };

  _processValue(value: string) {
    return flow([
      this._enforceStringValue,
      this._enforceMaxLength,
      this._enforceMinLength
    ]).call(this, value);
  }

  _enforceStringValue(value) {
    if (!isString(value)) {
      throw new Error('Values passed to Input::onChange must be strings');
    }
    return value;
  }

  _enforceMaxLength(value: string) {
    const { maxLength } = this.props;
    const isTooLong = maxLength != null && value.length > maxLength;
    return isTooLong ? value.substring(0, maxLength) : value;
  }

  _enforceMinLength(value: string) {
    const { minLength } = this.props;
    const isTooShort = minLength != null && value.length < minLength;

    if (isTooShort) {
      this._setError('Please enter a valid input');
    } else if (this.state.error !== '') {
      this._setError('');
    }

    return value;
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: InputSkin,
      context,
      theme,
      themeOverrides,
      onChange,
      error,
      maxLength,
      minLength,
      setError,
      autoFocus,
      ...rest
    } = this.props;

    return (
      <InputSkin
        error={error || this.state.error}
        onChange={this.onChange}
        inputRef={this.inputElement}
        theme={this.state.composedTheme}
        {...rest}
      />
    );
  }
}

export const Input = withTheme(InputBase);
