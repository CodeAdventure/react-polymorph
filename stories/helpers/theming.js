import React from 'react';
import { ThemeProvider } from '../../source/components/HOC/ThemeProvider';
import { SimpleTheme } from '../../source/themes/simple';

export const decorateWithSimpleTheme = (story) => (
  <ThemeProvider theme={SimpleTheme}>
    {story()}
  </ThemeProvider>
);
