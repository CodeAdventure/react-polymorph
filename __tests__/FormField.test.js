import React from 'react';

import { FormField } from '../source/components/FormField';
import { renderInSimpleTheme } from './helpers/theming';

const renderFormField = () => <div className="render-prop" />;

test('FormField renders correctly', () => {
  expect(renderInSimpleTheme(
    <FormField render={renderFormField} />
  )).toMatchSnapshot();
});

test('FormField renders with label', () => {
  expect(renderInSimpleTheme(
    <FormField
      label="Add a Label"
      render={renderFormField}
    />
  )).toMatchSnapshot();
});

test('FormField renders with an error', () => {
  expect(renderInSimpleTheme(
    <FormField
      error="Add an Error"
      render={renderFormField}
    />
  )).toMatchSnapshot();
});

test('FormField is disabled', () => {
  expect(renderInSimpleTheme(
    <FormField
      disabled
      render={({ disabled }) => <span>{disabled.toString()}</span>}
    />
  )).toMatchSnapshot();
});

test('FormField renders an input element', () => {
  expect(renderInSimpleTheme(
    <FormField
      render={() => <input className="render-prop" />}
    />
  )).toMatchSnapshot();
});
