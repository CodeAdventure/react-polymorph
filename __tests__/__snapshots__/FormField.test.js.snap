// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField is disabled 1`] = `
<div
  class="root disabled"
>
  <div
    class="inputWrapper"
  >
    <span>
      true
    </span>
  </div>
</div>
`;

exports[`FormField renders an input element 1`] = `
<div
  class="root"
>
  <div
    class="inputWrapper"
  >
    <input
      class="render-prop"
    />
  </div>
</div>
`;

exports[`FormField renders correctly 1`] = `
<div
  class="root"
>
  <div
    class="inputWrapper"
  >
    <div
      class="render-prop"
    />
  </div>
</div>
`;

exports[`FormField renders with an error 1`] = `
<div
  class="root errored"
>
  <div
    class="error"
  >
    Add an Error
  </div>
  <div
    class="inputWrapper"
  >
    <div
      class="render-prop"
    />
  </div>
</div>
`;

exports[`FormField renders with label 1`] = `
<div
  class="root"
>
  <label
    aria-hidden="true"
    class="label"
    role="presentation"
  >
    Add a Label
  </label>
  <div
    class="inputWrapper"
  >
    <div
      class="render-prop"
    />
  </div>
</div>
`;
